### PR TITLE
drivers: nrf: Enable nRF peripheral drivers by default for nRF SoCs

### LINKED ISSUE #11902: Nordic devices require extra configuration for I2C
> Ok, then it looks like the nordic parts don't handle this like I'd hoped. @Mierunski, @anangl you might want to consider doing this because it allows for fewer soc-specific .conf files in the sample and test applications.

https://github.com/zephyrproject-rtos/zephyr/blob/b5ae7da2697b6e38180856f014cfe6bbc652e4e6/soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk64f12#L46-L51

https://github.com/zephyrproject-rtos/zephyr/blob/b5ae7da2697b6e38180856f014cfe6bbc652e4e6/boards/arm/frdm_k64f/Kconfig.defconfig#L73-L81

_Originally posted by @MaureenHelm in [this review comment](
https://github.com/zephyrproject-rtos/zephyr/pull/11545#discussion_r237860462)_

### PR DESCRIPTION
**Pull request by anangl**
These changes make the following nRF peripheral drivers:
- ADC
- GPIO
- I2C
- SPI
- UART
- USB_DEVICE

enabled by default, _EDIT: when given feature is used in the application (i.e. enabled in Kconfig)_, so that users do not need to explicitly enable them in their applications after choosing an nRF SoC as the build target.  

Also default instances of ADC, I2C, and SPI hardware are enabled for nRF development kit boards that have the corresponding DT nodes enabled. The idea behind that is also to free users from the requirement to do it at the application level, so they can conveniently build for instance sensor samples.

After applying the above, the related Kconfig adjustments made on board level and also in samples and tests became redundant, hence they are removed in this PR.

Fixes #11902

On the occasion of these changes, two minor corrections in nRF drivers related Kconfig files are done:
- a false dependency of serial drivers on gpio driver is removed
- the default value for SPIM3 MISO sampling delay is set to 2 so that it is consistent with what appears in the hardware after reset

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]
LGTM thanks @anangl. I will let @pabigot and anyone else review.

**Comment by anangl** [comment]
Corrected the whitespace errors (spaces instead of tabs in _nrf51_ble400.dts_).

**Review by ioannisg — APPROVED** [review]

File: drivers/spi/Kconfig.nrfx [hunk 1]
```diff
@@ -216,7 +217,7 @@ config SPI_3_NRF_ORC
 config SPI_3_NRF_RX_DELAY
 	int "MISO sampling delay"
 	range 0 7
-	default 0
+	default 2
```
[the remark this answers is not in the document]
Comment (anangl) [reply drivers/spi/Kconfig.nrfx]: I added a mention of this change in the PR description.

**Review by Mierunski — APPROVED** [review]

**Review by pabigot — CHANGES_REQUESTED** [review]
Alternative to dropping the nrf51_ble400 changes, you could incorporate the contents of #11905 to that commit.

File: boards/arm/nrf51_ble400/nrf51_ble400.dts [hunk 2]
```diff
@@ -81,6 +81,12 @@
 	status ="ok";
 };
 
+&i2c0 {
```
[the remark this answers is not in the document]
Comment (anangl) [reply boards/arm/nrf51_ble400/nrf51_ble400.dts]: Ok, I'll drop it. I'll remove also the following change to _boards/arm/nrf51_ble400/Kconfig.defconfig_ as it is related:
```
diff --git a/boards/arm/nrf51_ble400/Kconfig.defconfig b/boards/arm/nrf51_ble400/Kconfig.defconfig
index 38ace52dcde..e306f69b173 100644
--- a/boards/arm/nrf51_ble400/Kconfig.defconfig
+++ b/boards/arm/nrf51_ble400/Kconfig.defconfig
@@ -9,4 +9,11 @@ if BOARD_NRF51_BLE400
 config BOARD
 	default "nrf51_ble400"
 
+if I2C
+
+config I2C_0
+	default y
+
+endif # I2C
+
 endif # BOARD_NRF51_BLE400
```
You can add this together with corresponding addition for SPI after applying #11905.

**Comment by anangl** [comment]
Dropped the changes related to I2C on nrf51_ble400, as requested by @pabigot.

**Review by pabigot — APPROVED** [review]

**Comment by aurel32** [comment]
> These changes make the following nRF peripheral drivers:
> 
>   - ADC
>   - GPIO
>   - I2C
>   - SPI
>   - UART
>   - USB_DEVICE
> 
> enabled by default so that users do not need to explicitly enable them in their applications after choosing an nRF SoC as the build target.

I understand the issue, but is it really the correct way to do that? It means that an application which does not use for example USB will still end-up with the USB driver compiled-in. This takes at least some flash for nothing.

Other SoC solves this issues by enabling the peripheral drivers in `Kconfig.defconfig.series` only if the feature is enable in Zephyr. For example on STM32:

```
if PWM

config PWM_STM32
        default y

endif # PWM

if SPI

config SPI_STM32
        default y

endif # SPI

if I2C

config I2C_STM32
        default y

endif

if USB

config USB_DC_STM32
        default y

endif # USB
```

On Kinetis K64F:
```

if ADC

config ADC_MCUX_ADC16
        default y

endif # ADC

if CLOCK_CONTROL

config CLOCK_CONTROL_MCUX_SIM
        default y

endif # CLOCK_CONTROL

if PINMUX

config PINMUX_MCUX
        default y

endif # PINMUX

config GPIO
        default y

if GPIO

config GPIO_MCUX
        default y

endif # GPIO

if I2C

config I2C_MCUX
        default y

endif # I2C

if PWM

config PWM_MCUX_FTM
        default y

endif # PWM

config SPI
        default n

if SPI

config SPI_MCUX_DSPI
        default y

endif # SPI

if NET_L2_ETHERNET

config ETH_MCUX
        default y

endif # NET_L2_ETHERNET
```

**Comment by pabigot** [comment]
@aurel32 I believe the implementation is exactly as you demonstrate, it's just that the wording of the commit message is a little subtle: it enables the *instances* by default.

In 8c7ae23b2e48b20d52c4e0d975771d5f04b73479 the change ensures that if `CONFIG_SPI` is on the default SPI peripheral instance for the board is also automatically enabled.

In d4e97fdc4e870c3d6f44aaae1a18fa0c064c41d8 it does the same thing for cases where there is only one peripheral instance, as with `CONFIG_ADC`.

In neither case is the driver automatically enabled.  If it is in some case, I agree that's worth a NACK.

**Review by erwango — CHANGES_REQUESTED** [review]
This is in contradiction with https://docs.zephyrproject.org/latest/porting/board_porting.html#default-board-configuration.
So this rule can be reviewed but it was agreed to make things more simple at application level since it makes clear to each application what is the baseline feature level.
So each application enables only the required set of features and don't get bloated with, say, unneeded USB driver.
As stated @aurel32 , there are ways to configure boards so  that application applications only need to enable high level features (USB, ..) without the need to go and check with driver or driver instance is required for each board.

File: drivers/usb/device/Kconfig [hunk 3]
```diff
@@ -58,6 +58,7 @@ config USB_DC_SAM
 
 config USB_NRF52840
 	bool "Nordic Semiconductor nRF52840 USB Device Controller Driver"
+	default y
```
Comment (finikorg) [inline drivers/usb/device/Kconfig]: This enables USB for all samples?

**Review by finikorg — CHANGES_REQUESTED** [review]
While you are here please also remove Bluetooth to be included in all samples, just built hello_world sample with Bluetooth enabled (for nrf52840_pca10059) :)

This patch d4e97fdc4e870c3d6f44aaae1a18fa0c064c41d8 looks OK.

**Comment by aurel32** [comment]
> @aurel32 I believe the implementation is exactly as you demonstrate, it's just that the wording of the commit message is a little subtle inaccurate for one, and subtle for the other (it enables the _instances_ by default)
> 
> In [8c7ae23](https://github.com/zephyrproject-rtos/zephyr/commit/8c7ae23b2e48b20d52c4e0d975771d5f04b73479) the change ensures that if `CONFIG_SPI` is on the default SPI peripheral instance for the board is also automatically enabled.

I just see a `default y` there, no dependency on `CONFIG_SPI`

> In [d4e97fd](https://github.com/zephyrproject-rtos/zephyr/commit/d4e97fdc4e870c3d6f44aaae1a18fa0c064c41d8) it does the same thing for cases where there is only one peripheral instance, as with `CONFIG_ADC`.

Indeed, that's exactly what should be done, but I think it should be done at the SoC level instead of being duplicated to each board.

**Comment by pabigot** [comment]
> Indeed, that's exactly what should be done, but I think it should be done at the SoC level instead of being duplicated to each board.

It can't, really; different boards use different instances in different ways.  The instances that are made available are on dedicated pins which are board-specific.

**Comment by pabigot** [comment]
@finikorg the automatic inclusion of bluetooth driver support in any application built on a Nordic boards is present in master, not from this patch.  I agree it should be removed; it's not obvious where it's coming from, and if removing it is complex it shouldn't block merge of this PR.

hello_world built with this patch set removes `gpio_nrfx` from the object files and adds nothing, so the effect of these changes is actually to reduce the number of unnecessary drivers.

**Comment by anangl** [comment]
@aurel32, @finikorg
> It means that an application which does not use for example USB will still end-up with the USB driver compiled-in. This takes at least some flash for nothing.

No, the USB driver will be compiled only when the USB option is activated in Kconfig, see [here](https://github.com/zephyrproject-rtos/zephyr/blob/8c7ae23b2e48b20d52c4e0d975771d5f04b73479/drivers/usb/device/Kconfig#L9).

As the 8c7ae23 commit message says later:

> Kconfig options enabling these drivers depend on both a given hardware
> feature (e.g. I2C) and an nRF family SoC selected, so effectively they
> will be automatically enabled only when it is adequate (and in most
> cases these drivers are the only option for a given hardware feature
> on nRF SoCs).

**Comment by anangl** [comment]
@aurel32

> > In [8c7ae23](https://github.com/zephyrproject-rtos/zephyr/commit/8c7ae23b2e48b20d52c4e0d975771d5f04b73479) the change ensures that if `CONFIG_SPI` is on the default SPI peripheral instance for the board is also automatically enabled.

> I just see a `default y` there, no dependency on `CONFIG_SPI`

The dependency is not achieved by `depend` in the `_NRFX` option definition but by the `if` in which the option is contained. See for instance SPI: the [`SPI_NRFX`](https://github.com/zephyrproject-rtos/zephyr/blob/8c7ae23b2e48b20d52c4e0d975771d5f04b73479/drivers/spi/Kconfig.nrfx#L6) option will be available only when the support for SPI is enabled, i.e. when the [`SPI`](https://github.com/zephyrproject-rtos/zephyr/blob/8c7ae23b2e48b20d52c4e0d975771d5f04b73479/drivers/spi/Kconfig#L12) option is enabled. This is because the _drivers/spi/Kconfig.nrfx_ file is [included](https://github.com/zephyrproject-rtos/zephyr/blob/8c7ae23b2e48b20d52c4e0d975771d5f04b73479/drivers/spi/Kconfig#L183) from inside [this `if`](https://github.com/zephyrproject-rtos/zephyr/blob/8c7ae23b2e48b20d52c4e0d975771d5f04b73479/drivers/spi/Kconfig#L18).

**Comment by finikorg** [comment]
> @aurel32, @finikorg
> 
> > It means that an application which does not use for example USB will still end-up with the USB driver compiled-in. This takes at least some flash for nothing.
> 
> No, the USB driver will be compiled only when the USB option is activated in Kconfig, see [here](https://github.com/zephyrproject-rtos/zephyr/blob/8c7ae23b2e48b20d52c4e0d975771d5f04b73479/drivers/usb/device/Kconfig#L9).
> 

So for other USB drivers it is selected i other places, could you do the same way to be consistent?

**Comment by pabigot** [comment]
> So for other USB drivers it is selected i other places, could you do the same way to be consistent?

isn't this the same place the other USB drivers are selected?

**Comment by pabigot** [comment]
> No, other drivers are not selected with `default y`

The others are done in board-specific defconfig, which was already the case for nRF52840-based boards.  Agreed, `default y` is not needed here.

**Comment by anangl** [comment]
@aurel32 
I updated the PR description. Now I see it was misleading earlier. Hopefully, now it clearly states what I'm trying to achieve with these changes. Apologies for not being precise enough from the beginning.

**Review by erwango — APPROVED** [review]
Updating my review.
Change is actually in line with default board configuration guidelines.

I'd propose to use following wording to avoid misunderstanding:
Boards features are **configured** by default.

Btw: fixes https://github.com/zephyrproject-rtos/zephyr/issues/9242

**Review by finikorg — APPROVED** [review]

**Comment by aurel32** [comment]
@anangl Thanks for the explanations and sorry about the confusion. Your solution of enabling the driver by default is actually smarter than what is currently done for other SoCs. It requires that the driver correctly depends on the SoC, which is not always the case for all drivers, but seems to be fine for nRF ones.

**Review by pabigot — CHANGES_REQUESTED** [review]
Sorry to continue the churn, but per @finikorg comment since this is a general cleanup please remove this from each board defconfig:
```
# bluetooth
CONFIG_BT=y
CONFIG_BT_CTLR=y
```

Saves over 30 KB in each image that doesn't use bluetooth.

**Comment by mike-scott** [comment]
@pabigot +1

And we start seeing BT=n / BT_CTRL=n in many of the .conf files when another connectivity option is needed (802.15.4 / OpenThread)

**Comment by anangl** [comment]
@pabigot @mike-scott
I agree with you. To me it also seems reasonable to not include bluetooth by default. But I don't really know why the boards were configured this way (perhaps some of them, and then maybe these settings just got propagated) and I can't take care of investigating this now and correcting the configurations (I suspect that this may require adding these settings in some applications). Besides, it's not really related with the topic of this PR. I disagree that this is a general cleanup what I'm doing here.
Let's solve this problem in another PR.

**Review by pabigot — APPROVED** [review]

**Comment by anangl** [comment]
@pabigot Thanks!

**Comment by pabigot** [comment]
Fair enough; that was my original assessment, but it does need to be fixed.  I'll open another issue for that.

**Comment by anangl** [comment]
recheck

**Pull request merged**